index.js

```diff
@@ -467,21 +472,30 @@ class LedgerBridgeKeyring extends EventEmitter {
 
   _sendMessage (msg, cb) {
     msg.target = 'LEDGER-IFRAME'
+
+    this.currentMessageId++
+    msg.messageId = this.currentMessageId
+
+    this.messageCallbacks[this.currentMessageId] = cb
     this.iframe.contentWindow.postMessage(msg, '*')
+  }
+
+  _setupListener () {
     const eventListener = ({ origin, data }) => {
       if (origin !== this._getOrigin()) {
         return false
       }
 
-      if (data && data.action && data.action === `${msg.action}-reply` && cb) {
-        cb(data)
-        return undefined
+      if (data && this.messageCallbacks[data.messageId]) {
+        this.messageCallbacks[data.messageId](data)
       }
 
-      window.removeEventListener('message', eventListener)
       return undefined
     }
     window.addEventListener('message', eventListener)
+    window.addEventListener('beforeunload', () => {
```

Comment:
It doesn't seem entirely necessary to remove the event listener before unload. It'll get dropped on unload anyway.